biggraphite/glob_utils.py

```diff
@@ -0,0 +1,204 @@
+                res = res + '\\['
+            else:
+                tmp = glob[i:j].replace('\\', '\\\\')
+                i = j+1
+                if tmp[0] == '!':
+                    tmp = '^' + tmp[1:]
+                elif tmp[0] == '^':
+                    tmp = '\\' + tmp
+                res += '[' + tmp + ']'
+        # Word selection (supports nesting)
+        elif c == '{':
+            depth = 1
+            tmp = '('
+            j = i
+            # Look-ahead for the whole brace-enclosed expression
+            while j < n and depth > 0:
+                c = glob[j]
+                j += 1
+                if c == '{':
+                    tmp += '('
+                    depth += 1
+                elif c == '}':
+                    tmp += ')'
+                    depth -= 1
+                elif c == ',':
+                    tmp += '|'
+                else:
+                    tmp += c
+            # Escape the current brace in case it is mismatched
+            if depth > 0:
+                res += '\\{'
+            else:
+                i = j
+                res += tmp
+        # Default: escape character
+        else:
+            res = res + re.escape(c)
+    return '^' + res + '$'
+
+
+def glob_prefilter(metric_names, graphite_glob):
+    """Pre-filter metric names according to a glob expression.
+
+    Uses the dot-count and the litteral components of the glob to filter
+    guaranteed non-matching values out, but may still require post-filtering.
+
+    Args:
+      metric_names: List of metric names to be filtered.
+      graphite_glob: Graphite glob pattern.
+
+    Returns:
+      List of metric names that may be matched by the provided glob.
+    """
+    dots = graphite_glob.count(".")
+    literal_glob_components = [
+        (index, component)
+        for (index, component) in enumerate(graphite_glob.split("."))
+        if not _is_graphite_glob(component)
+    ]
+
+    def maybe_matched_prefilter(metric):
+        if metric.count(".") != dots:
+            return False
+        metric_components = metric.split(".")
+        for (index, value) in literal_glob_components:
+            if metric_components[index] != value:
+                return False
+        return True
+
+    return filter(maybe_matched_prefilter, metric_names)
+
+
+def glob(accessor, graphite_glob):
+    """Get Cassandra metrics and directories matching a Graphite glob.
+
+    Args:
+      accessor: Cassandra accessor
+      graphite_glob: Graphite glob expression
+
+    Returns:
+      A tuple:
+        First element: sorted list of Cassandra metrics matched by the glob.
+        Second element: sorted list of Cassandra directories matched by the glob.
+    """
+    if not _is_valid_glob(graphite_glob):
+        # TODO(d.forest): should we instead raise an exception?
+        return ([], [])
+
+    glob_re = re.compile(_glob_to_regex(graphite_glob))
+    accessor_components = _graphite_glob_to_accessor_components(graphite_glob)
```

Comment:
I think there still a misunderstanding, in my mind we should have:
* graphite_glob(), used only by the graphite plugin: supports graphite thingies.
* glob(), used by the drivers when the database can't do it: supports star and globstar

Graphite specific things should *stay* in the graphite plugin scope and not leak further.